src/login/logind.c

```diff
@@ -409,10 +409,72 @@ static int manager_enumerate_users(Manager *m) {
         return r;
 }
 
+static int manager_attach_fds(Manager *m) {
+        _cleanup_strv_free_ char **fdnames = NULL;
+        int r = 0, n, i, fd;
+
+        /* Upon restart, PID1 will send us back all fds of session devices
+         * that we previously opened. Each file descriptor is associated
+         * with a given session. The session ids are passed through FDNAMES. */
+
+        n = sd_listen_fds_with_names(true, &fdnames);
+        if (n <= 0)
+                return n;
+
+        for (i = 0; i < n; i++) {
+                struct stat st;
+                SessionDevice *sd;
+                Session *s;
+                char *id;
+
+                fd = SD_LISTEN_FDS_START + i;
+
+                if (!startswith(fdnames[i], "session-"))
+                        continue;
+
+                id = fdnames[i] + 8;
```

Comment:
not that it matters, but the three lines above could be rewritten in a slightly less opaque way as:

```
id = startswith(fdnames[i], "session-"));
if (!id)
        continue;
```
i.e. taking benefit of the fact that startswith() returns a pointer to the first character after the determined prefix